### DISCUSSION TITLE
Republish apps and not publish apps

### DISCUSSION THREAD

**Discussion by HUJON1234** [ptarmiganlabs/ctrl-q#234] [Ideas]
Hello!

I would love to have an option to republish for an app. I would like to have:
 App Name - App stream as unique key if it already exists then overwrite (republish) the app.

I really hope this is possible in the near future.

**Comment by mountaindude** [comment]
Good idea. 

Just so I understand correctly: 
This would be when importing apps defined in the Excel file, i.e. the `task-import` command. 
Similar to what's described [here](https://ptarmiganlabs.com/the-lazy-qlik-developer-3-bulk-export-import-of-both-apps-and-reload-tasks/), but re-publishing apps?

**Comment by HUJON1234** [comment]
Yes! 

That is exactly what i want. 

for example we have on server 1:
AppName 1 and StreamName 1

and on server 2: the apps already exist
AppName 1 and StreamName 1

then the app has to re publish an app or delete the app on server 2.

**Comment by mountaindude** [comment]
Having given this some thought I can see several possible scenarios for doing upload-app-republish.

My current thinking is to add a new column in the source Excel file.
This column would specify what to do with the uploaded QVF file/app.
For example, one value could be `replace-if-exists-in-stream`, another could be `delete-existing-in-stream-then-publish`.
And so on. The exact values in that column may turn out to be something else than above, but you hopefully get the point.

For your use case the `replace-if-exists-in-stream` would be good I think.
It's exactly what happens when you do a publish-replace from the QMC, but with a source file that has just been uploaded from QVF on disk.

One question is how to handle the app that was created from the QVF (i.e. the app that is the source of the publish-replace operation).
I think that app should be deleted once the publish-replace operation is done. 
If not deleted we would get a strange situation down the road, when running the app import command next time (we'd get many just-uploaded, unpublished apps).
I think the just uploaded app should be viewed as temporary and as such deleted when the publish-replace is done.

Thoughts on your end?

**Reply by HUJON1234** [reply]
Yes!

I totally agree, for my use case replace-if-exists-in-stream is the way to go. 

Yes, i thought i would write another script to delete the files, if you make it work in the tool it would be the best option.

**Comment by mountaindude** [comment]
Design notes

- A new column "Publish options" added to the Excel sheet where apps to be imported are listed. Possible values will be
  - `publish-replace`: 
    1. Upload the app specified in the Excel file.
    2. Check how many already published apps exists with the same name in the target stream
       - If 0: Do a normal publish
       - If 1: Do a publish-replace, exactly the same as in the QMC. Delete the uploaded app to avoid getting lots of unpublished/temporary apps.
       - If 2 or more: It is impossible to know which of the published apps to replace. Warn in the logs and skip publishing this app.
  - `publish-another`:
    1. Upload the app specified in the Excel file.
    4. Publish the uploaded app into the target stream, even if an app with that name already exists in that stream. 
    5. The result is that there may be several apps with the same name (but different app IDs) in a stream. 
    6. This is the behaviour in Ctrl-Q version 3.11 and earlier.
  - `delete-publish`: 
    1. Upload the app specified in the Excel file.
    2. If an app with the same name as uploaded file already exists in the target stream: Delete that app.
    4. Publish the uploaded app. The published app will get the same app ID as the app that was uploaded.
- Make the new "Publish options" column optional. I.e. the old Excel file format should still work.
  - The default behaviour is `publish-replace ` and will be used when
    - No "Publish options" column is present in the Excel sheet.
    - The column exists but is empty. 
    **NOTE**: This will be changed behaviour to the current behaviour, which is `publish-another`

**Comment by mountaindude** [comment]
Version 3.14 was just released (download [here](https://github.com/ptarmiganlabs/ctrl-q/releases/tag/v3.14.0)), it contains (among other things) changes related to this ticket.
